### PR TITLE
[WIP] Presto udf post

### PR DESCRIPTION
**Pull request by dpawlow-jampp**
TODO:

- [ ] Reference post in indexes.
- [ ] Copy UDF to a public repo.
- [ ] Reference links to the above mentioned repo.
- [ ] Conclusion.
- [ ] Pictures?
- [ ] Editor review.

### CODE REVIEW

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+Presto comes with many [functions](https://prestodb.io/docs/current/functions.html), which are usually enough for most use cases. Nevertheless, sometimes you need to implement your own function for a very specific use.
+
+Enter the [User Defined Functions](https://prestodb.io/docs/current/develop/functions.html) (UDFs, for short). Writing one is not as straightforward as it first appears, mainly because the information to do so is very scattered around the web (and across many Presto versions).
+
+In this blogpost, we present our _JSON Sum_ function, how did we write it, and some of the lessons we learned along the way.
+<!--excerpt.end-->
+
+## Function types
+
+Presto has two main types of functions: scalar and aggregation[^1]. Scalar functions behave more like traditional functions, in the sense that their output is immediate from the values of their parameters. On the other hand, aggregation functions take multiple rows as input and combine them into a single output, we'll focus mainly on these functions, as they're more complex (and, interesting!).
```
Comment (eorbe) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: `... a single output, we'll focus mainly on these functions, as they're more complex (and, interesting!).`

I would rephrase it as:

`... a single output. We'll focus mainly on the latter, as they're more complex (and, interesting!).`

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+[^1]: There's also window functions, but we have yet to implement one of those, so we'll leave them out of the scope of this article.
+
+Aggregation functions can harness the power of presto’s distributed workers via a divide and conquer approach.
+They consist of three main methods:
+
+    - _input_, which reads each row and feeds it to a state.
+    - _combine_, invoked whenever two states need to be merged.
+    - _output_, which returns the result of the aggregation at the end.
+
+The _state_ is a data structure that stores the meaningful results of each step of the aggregation, and can be designed according to the specific needs. It should be serializable (and deserializable) to share it between nodes.
```
Comment (eorbe) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: `... according to the specific needs. It should be serializable (and deserializable) to share it between nodes.`

I would rephrase it as:

`... according to the specific needs, the only requirement being for it to be serializable in order to be shared between the workers.`

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+It's important to note that no order of input or combination should be assumed: any given state should be able to be combined with any other and any row should be able to be converted to a valid state. You can, however, define the initial parameters of the state via its constructor.
+
+## Implementation
+
+UDFs are implemented inside a [plugin](https://prestodb.io/docs/current/develop/spi-overview.html), separate from the default Presto functions. They are exposed to presto by declaring them in [UdfPlugin.java] and later deployed as .jar files in your cluster.
```
Comment (eorbe) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: `by declaring them in [UdfPlugin.java]` -> `by declaring them in the [UdfPlugin.java] file and then deploying them as a .jar file`

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        return ImmutableSet.<Class<?>>builder()
+                .add(JSONAggregation.class)
+                .add(YourUDF.class)
+                .build();
+    }
+}
+```
+
+Aggregation functions are trickier to implement than scalar functions, as they have many more moving parts. Aside from the _input_, _combine_ and _output_ functions, you should write a [State](TODO) interface and its auxiliary files. If your state uses just basic data types, Presto automatically knows how to construct, serialize and deserialize it. Else, you should implement a [Factory](TODO) and a [Serializer](TODO), and link them to the State with Presto's Metadata Annotations.
```
Comment (eorbe) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: `...State with Presto's Metadata Annotations.` -> `...State using Presto's Metadata Annotations.`

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+Once it's up, it can be queried like any other Presto cluster, so you can use it to test your UDF.
+
+## Pitfalls
+
+> _"My kingdom for a discernible stack trace!"_
+>> King Richard III
+
+Developing software never is a smooth-sailing kind of deal (where would the fun in that be?), so be prepared to stare into weird looking stack traces. Here are some of the traps we falled into, hopefully we'll save you some headaches.
+
+* Do NOT use the presto-main dependency for anything besides testing (and even then, try to avoid it). Whichever function you need from presto-main most likely has an equivalent in the spi-package.
```
Comment (eorbe) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Probably a link to some reference about `presto-main` and the  `spi-package` would be useful.

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown (hunk 2 above)
Comment (dpawlow-jampp) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Maybe `...specific needs, the only requirement being that it's serializable to be shareable between Presto's workers.` ?

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown (hunk 1 above)
Comment (dpawlow-jampp) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Good suggestion!

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    - technology
+keywords: "Presto, PrestoDB, UDF, aggregation"
+author: dantepawlow
+---
+
+# Writing custom PrestoDB functions
+
+<!--excerpt.start-->
+
+Here at Jampp we process and analyse large amounts of data.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *analyze

for American English

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+One of the tools we employ to do so is [PrestoDB](https://prestodb.io/), which is a "Distributed SQL Query Engine for Big Data".
+Presto comes with many [native functions](https://prestodb.io/docs/current/functions.html), which are usually enough for most use cases. Nevertheless, sometimes you need to implement your own function for a very specific use.
+
+Enter the [User Defined Functions](https://prestodb.io/docs/current/develop/functions.html) (UDFs, for short).
+Writing one is not as straightforward as it first appears, mainly because the information to do so is very scattered around the web (and across many Presto versions).
+
+In this blogpost, we present our *JSON_Sum* function, how did we write it, and some of the lessons we learned along the way.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: SUM all caps

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+# Writing custom PrestoDB functions
+
+<!--excerpt.start-->
+
+Here at Jampp we process and analyse large amounts of data.
+One of the tools we employ to do so is [PrestoDB](https://prestodb.io/), which is a "Distributed SQL Query Engine for Big Data".
+Presto comes with many [native functions](https://prestodb.io/docs/current/functions.html), which are usually enough for most use cases. Nevertheless, sometimes you need to implement your own function for a very specific use.
+
+Enter the [User Defined Functions](https://prestodb.io/docs/current/develop/functions.html) (UDFs, for short).
+Writing one is not as straightforward as it first appears, mainly because the information to do so is very scattered around the web (and across many Presto versions).
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Explain that UDFs are custom made functions in Java

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+In this blogpost, we present our *JSON_Sum* function, how did we write it, and some of the lessons we learned along the way.
+
+<!--excerpt.end-->
+
+## Function types
+
+Presto has two main types of functions: scalar and aggregation[^1].
+Scalar functions behave more like traditional functions, in the sense that their output is immediate from the values of their parameters.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Thinks some alternative for immediate.  I don't think it's really clear what you are trying to say.

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+On the other hand, aggregation functions take multiple rows as input and combine them into a single output.
+We'll focus mainly on the latter, as they're more complex (and, interesting!).
+
+[^1]: There's also window functions, but we have yet to implement one of those, so we'll leave them out of the scope of this article.
+
+Aggregation functions can harness the power of presto’s distributed workers via a divide and conquer approach.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *Presto

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+The `State` is a data structure that stores the meaningful results of each step of the aggregation, and can be designed according to the specific needs, the only requirement being that it's serializable to be shareable between workers.
+
+It's important to note that no order of input or combination should be assumed: any given state should be able to be combined with any other and any row should be able to be converted to a valid state.
+You can, however, define the initial parameters of the state via its constructor.
+
+## Implementation
+
+UDFs are implemented inside a [plugin](https://prestodb.io/docs/current/develop/spi-overview.html), separate from the default Presto functions.
+They are exposed to presto by declaring them in [UdfPlugin.java](TODO) and then deploying them as a `.jar` file.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *Presto

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+[^2]: Thanks to [Lewuathe's docker-presto-cluster](https://github.com/Lewuathe/docker-presto-cluster)!
+
+Once it's up, it can be queried like any other Presto cluster, so you can use it to test your UDF.
+
+## Pitfalls
+
+> _"My kingdom for a discernible stack trace!"_
+>> King Richard III
+
+Developing software never is a smooth-sailing kind of deal (where would the fun in that be?), so be prepared to stare into weird looking stack traces.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Is it into or at?

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+Here are some of the traps we falled into, hopefully we'll save you some headaches.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *fallen

Comment (juanpampliega) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: we'll -> they will

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown (hunk 9 above)
Comment (dpawlow-jampp) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: "Scalar functions behave more like traditional functions, in the sense that their output is intuitively derived from their parameters."?

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown (hunk 12 above)
Comment (dpawlow-jampp) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: They're slightly different actions, I would argue that "stare into" fits better when you're trying to make sense of a stack trace: https://english.stackexchange.com/a/120603

Comment (juanpampliega) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: From my point of view you "stare at a screen" so you are "staring at a stack trace". But as you prefer.

Comment (santijampp) [reply _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Into adds a felling of deepness, what I imagine from a weird stack trace that goes on and on into function calls. :man_shrugging:

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+Developing software never is a smooth-sailing kind of deal (where would the fun in that be?), so be prepared to stare into weird looking stack traces.
+Here are some of the traps we've fallen into, hopefully they'll save you some headaches:
+
+* Do NOT use the [presto-main](https://mvnrepository.com/artifact/com.facebook.presto/presto-main) dependency for anything besides testing (and even then, try to avoid it). Whichever function you need from presto-main most likely has an equivalent in the [presto-spi](https://mvnrepository.com/artifact/com.facebook.presto/presto-spi) package.
+* If your state uses more complex data types than the basic types, you should add a `Factory` and a `Serializer` (else, you get weird errors like "HashMap<> not supported").
+* Always check that the `.jar` files are deployed in both the coordinator and the workers. If you get an esoteric Presto exception (like “varchar not found” even though the function is listed in the `SHOW FUNCTIONS` query), this is the most likely suspect.
+
+## Conclusion
+
+Pesto UDFs are a very powerful tool, but given the wide array of native functions at your disposal, it's likely that you'll never need to write one.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *Presto

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        return ImmutableSet.<Class<?>>builder()
+                .add(JSONAggregation.class)
+                .add(YourUDF.class)
+                .build();
+    }
+}
+```
+
+**Scalar functions** are fairly simple: you write a single java function, and annotate it appropriately.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: *Java

File: _src/_posts/2018-09-07-writing-custom-presto-functions.markdown [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+This works in our case, but if you need your UDF to be more generic, you can follow PrestoDB's example in their [native functions](https://github.com/prestodb/presto/tree/master/presto-main/src/main/java/com/facebook/presto/operator/aggregation).
+
+## Testing
+
+Once you have your UDF, the sensible thing is to test it. But how do you simulate Presto's inner workings? This is especially important for aggregation functions, given that Presto takes the components of your UDF and uses them inside a black box.
+Thankfully, we have [Docker](https://www.docker.com/)! [Here](TODO) we provide a Makefile and a docker-compose.yml[^2] for getting a Presto server up and running locally, automatically incorporating your UDF.
+
+[^2]: Thanks to [Lewuathe's docker-presto-cluster](https://github.com/Lewuathe/docker-presto-cluster)!
+
+Once it's up, it can be queried like any other Presto cluster, so you can use it to test your UDF.
```
Comment (juanpampliega) [inline _src/_posts/2018-09-07-writing-custom-presto-functions.markdown]: Maybe add the query example of how be tested this function.
